From: Ray Strode <rstrode@redhat.com>
Date: Tue, 6 Feb 2024 11:13:31 -0500
Subject: animation: Allow spinner to be sizes other than 16x16 pixels

Since sliced images are provided in a two dimensional grid, instead
of linear strip, there's no way to infer the size of an individual
cell inside a sliced image. Because of this limitation the texture
cache api for loading an image requires the cell size (called grid
size in the api) be explicitly specified. But because sliced images
are scalable SVGs, the texture cache code needs to know the output
animation size as well, to ensure the animation is rendered with an
appropriately high resolution. The API doesn't ask for this
information, however, instead just assuming the output size is the same
as the cell size. That effectively means that spinners can only be
16x16 pixels big (because our spinner slicable images have cells that
are nominally 16x16).

This commit changes the API to allow the output image size to be
specified as well, and fixes the code to render at the appropriate
output size.
---
 js/ui/animation.js        | 65 +++++++++++++++++++++++++++++++++++++++--------
 src/st/st-texture-cache.c | 14 +++++++---
 src/st/st-texture-cache.h |  2 ++
 3 files changed, 67 insertions(+), 14 deletions(-)

diff --git a/js/ui/animation.js b/js/ui/animation.js
index feffc19..a0b8de9 100644
--- a/js/ui/animation.js
+++ b/js/ui/animation.js
@@ -9,19 +9,52 @@ import * as Params from '../misc/params.js';
 const ANIMATED_ICON_UPDATE_TIMEOUT = 16;
 const SPINNER_ANIMATION_TIME = 300;
 const SPINNER_ANIMATION_DELAY = 1000;
+const SPINNER_ANIMATION_CELL_SIZE = 16;
 
 export const Animation = GObject.registerClass(
 class Animation extends St.Bin {
-    _init(file, width, height, speed) {
+    _init(params) {
         const themeContext = St.ThemeContext.get_for_stage(global.stage);
+        params = Params.parse(params, {
+            file: null,
+            cellWidth: null,
+            cellHeight: null,
+            frameWidth: null,
+            frameHeight: null,
+            speed: 0,
+        });
+
+        const {file, speed} = params;
+        let {frameWidth, frameHeight, cellHeight, cellWidth} = params;
+
+        if (!cellWidth)
+            throw new Error('Animation has unspecified cell width');
+
+        if (!frameHeight) {
+            if (frameWidth)
+                frameHeight = frameWidth;
+            else
+                frameHeight = cellHeight;
+        }
+
+        if (!frameWidth)
+            frameWidth = cellWidth;
+
+        if (!cellHeight)
+            cellHeight = cellWidth;
+
+        if (!frameHeight)
+            frameHeight = cellWidth;
 
         super._init({
-            style: `width: ${width}px; height: ${height}px;`,
+            style: `width: ${frameWidth}px; height: ${frameHeight}px;`,
         });
 
         this._file = file;
-        this._width = width;
-        this._height = height;
+        this._cellWidth = cellWidth;
+        this._cellHeight = cellHeight;
+        this._frameWidth = frameWidth;
+        this._frameHeight = frameHeight;
 
         this.connect('destroy', this._onDestroy.bind(this));
         this.connect('resource-scale-changed',
@@ -31,8 +64,8 @@ class Animation extends St.Bin {
             () => {
                 this._loadFile();
                 this.set_size(
-                    this._width * themeContext.scale_factor,
-                    this._height * themeContext.scale_factor);
+                    this._frameWidth * themeContext.scale_factor,
+                    this._frameHeight * themeContext.scale_factor);
             }, this);
 
         this._speed = speed;
@@ -79,7 +112,8 @@ class Animation extends St.Bin {
         let textureCache = St.TextureCache.get_default();
         let scaleFactor = St.ThemeContext.get_for_stage(global.stage).scale_factor;
         this._animations = textureCache.load_sliced_image(this._file,
-            this._width, this._height,
+            this._cellWidth, this._cellHeight,
+            this._frameWidth, this._frameHeight,
             scaleFactor, resourceScale,
             () => this._loadFinished());
         this._animations.set({
@@ -124,8 +158,15 @@ class Animation extends St.Bin {
 
 export const AnimatedIcon = GObject.registerClass(
 class AnimatedIcon extends Animation {
-    _init(file, size) {
-        super._init(file, size, size, ANIMATED_ICON_UPDATE_TIMEOUT);
+    _init(params) {
+        const {file, cellSize, frameSize} = params;
+
+        super._init({
+            file,
+            cellWidth: cellSize,
+            frameWidth: frameSize,
+            speed: ANIMATED_ICON_UPDATE_TIMEOUT,
+        });
     }
 });
 
@@ -140,7 +181,11 @@ class Spinner extends AnimatedIcon {
             'resource:///org/gnome/shell/theme/process-working-dark.svg');
         this._fileLight = Gio.File.new_for_uri(
             'resource:///org/gnome/shell/theme/process-working-light.svg');
-        super._init(this._fileDark, size);
+        super._init({
+            file: this._fileDark,
+            cellSize: SPINNER_ANIMATION_CELL_SIZE,
+            frameSize: size,
+        });
 
         this.connect('style-changed', () => {
             const themeNode = this.get_theme_node();
diff --git a/src/st/st-texture-cache.c b/src/st/st-texture-cache.c
index a137480..74b0677 100644
--- a/src/st/st-texture-cache.c
+++ b/src/st/st-texture-cache.c
@@ -1198,6 +1198,7 @@ ensure_monitor_for_file (StTextureCache *cache,
 typedef struct {
   GFile *gfile;
   gint   grid_width, grid_height;
+  gfloat output_scale;
   gint   paint_scale;
   gfloat resource_scale;
   ClutterActor *actor;
@@ -1283,7 +1284,7 @@ on_loader_size_prepared (GdkPixbufLoader *loader,
                          gpointer user_data)
 {
   AsyncImageData *data = user_data;
-  int scale = ceilf (data->paint_scale * data->resource_scale);
+  int scale = ceilf (data->paint_scale * data->resource_scale * data->output_scale);
 
   gdk_pixbuf_loader_set_size (loader, width * scale, height * scale);
 }
@@ -1330,7 +1331,7 @@ load_sliced_image (GTask        *result,
   pix = gdk_pixbuf_loader_get_pixbuf (loader);
   width = gdk_pixbuf_get_width (pix);
   height = gdk_pixbuf_get_height (pix);
-  scale_factor = ceilf (data->paint_scale * data->resource_scale);
+  scale_factor = ceilf (data->paint_scale * data->resource_scale * data->output_scale);
   for (y = 0; y < height; y += data->grid_height * scale_factor)
     {
       for (x = 0; x < width; x += data->grid_width * scale_factor)
@@ -1356,8 +1357,10 @@ load_sliced_image (GTask        *result,
  * st_texture_cache_load_sliced_image:
  * @cache: A #StTextureCache
  * @file: A #GFile
- * @grid_width: Width in pixels
- * @grid_height: Height in pixels
+ * @grid_width: Width of animation cell in pixels
+ * @grid_height: Height of animation cell in pixels
+ * @image_width: Width of output image in pixels
+ * @image_height: Height of output image in pixels
  * @paint_scale: Scale factor of the display
  * @load_callback: (scope async) (nullable): Function called when the image is loaded, or %NULL
  * @user_data: Data to pass to the load callback
@@ -1374,6 +1377,8 @@ st_texture_cache_load_sliced_image (StTextureCache *cache,
                                     GFile          *file,
                                     gint            grid_width,
                                     gint            grid_height,
+                                    gint            image_width,
+                                    gint            image_height,
                                     gint            paint_scale,
                                     gfloat          resource_scale,
                                     GFunc           load_callback,
@@ -1391,6 +1396,7 @@ st_texture_cache_load_sliced_image (StTextureCache *cache,
   data = g_new0 (AsyncImageData, 1);
   data->grid_width = grid_width;
   data->grid_height = grid_height;
+  data->output_scale = MIN((double) image_width / grid_width, (double) image_height / grid_height);
   data->paint_scale = paint_scale;
   data->resource_scale = resource_scale;
   data->gfile = g_object_ref (file);
diff --git a/src/st/st-texture-cache.h b/src/st/st-texture-cache.h
index 1c5be95..05da456 100644
--- a/src/st/st-texture-cache.h
+++ b/src/st/st-texture-cache.h
@@ -58,6 +58,8 @@ st_texture_cache_load_sliced_image (StTextureCache *cache,
                                     GFile          *file,
                                     gint            grid_width,
                                     gint            grid_height,
+                                    gint            image_width,
+                                    gint            image_height,
                                     gint            paint_scale,
                                     gfloat          resource_scale,
                                     GFunc           load_callback,
